File: generated-code/php/example/FileReadWrite/OneOf.php
<?php

namespace  {
    require_once 'Stream.php';

    /**
     * Oneof example
     */
    abstract class OneOf
    {
        /**
         * Write OneOf to output stream
         */
        abstract function writeTo(\OutputStream $stream): void;

        /**
         * Read OneOf from input stream
         */
        static function readFrom(\InputStream $stream): OneOf
        {
            $tag = $stream->readInt32();
            if ($tag == \OneOf\OptionOne::TAG) {
                return \OneOf\OptionOne::readFrom($stream);
            }
            if ($tag == \OneOf\OptionTwo::TAG) {
                return \OneOf\OptionTwo::readFrom($stream);
            }
            throw new Exception('Unexpected tag value');
        }
    }
}

namespace OneOf {
    /**
     * First option
     */
    class OptionOne extends \OneOf
    {
        const TAG = 0;
    
        /**
         * List of integers
         */
        public array $vecInt;
        /**
         * Long integer
         */
        public int $longInt;
    
        function __construct(array $vecInt, int $longInt)
        {
            $this->vecInt = $vecInt;
            $this->longInt = $longInt;
        }
    
        /**
         * Read OptionOne from input stream
         */
        public static function readFrom(\InputStream $stream): OptionOne
        {
            $vecInt = [];
            $vecIntSize = $stream->readInt32();
            for ($vecIntIndex = 0; $vecIntIndex < $vecIntSize; $vecIntIndex++) {
                $vecIntElement = $stream->readInt32();
                $vecInt[] = $vecIntElement;
            }
            $longInt = $stream->readInt64();
            return new OptionOne($vecInt, $longInt);
        }
        
        /**
         * Write OptionOne to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(OptionOne::TAG);
            $stream->writeInt32(count($this->vecInt));
            foreach ($this->vecInt as $element) {
                $stream->writeInt32($element);
            }
            $stream->writeInt64($this->longInt);
        }
    }

    /**
     * Second option
     */
    class OptionTwo extends \OneOf
    {
        const TAG = 1;
    
        /**
         * usize
         */
        public int $value;
    
        function __construct(int $value)
        {
            $this->value = $value;
        }
    
        /**
         * Read OptionTwo from input stream
         */
        public static function readFrom(\InputStream $stream): OptionTwo
        {
            $value = $stream->readInt32();
            return new OptionTwo($value);
        }
        
        /**
         * Write OptionTwo to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(OptionTwo::TAG);
            $stream->writeInt32($this->value);
        }
    }
}

File: generated-code/pascal/example/FileReadWrite/OneOf.pas
unit OneOf;

interface

uses
    Stream,
    SysUtils;

type
    // Oneof example
    TOneOf = class
        // Write OneOf to output stream
        procedure WriteTo(stream: TStream); virtual; abstract;
        // Read OneOf from input stream
        class function ReadFrom(stream: TStream): TOneOf; static;
    end;

type
    // First option
    TOneOfOptionOne = class (TOneOf)
        // List of integers
        vecInt32: TArray<Int32>;
        // Long integer
        longInt: Int64;
        constructor Create(vecInt32: TArray<Int32>; longInt: Int64);
        // Read OneOfOptionOne from input stream
        class function ReadFrom(stream: TStream): TOneOfOptionOne; static;
        // Write OneOfOptionOne to output stream
        procedure WriteTo(stream: TStream); override;
        function ToString: ansistring; override;
    end;

type
    // Second option
    TOneOfOptionTwo = class (TOneOf)
        // usize
        value: Int32;
        constructor Create(value: Int32);
        // Read OneOfOptionTwo from input stream
        class function ReadFrom(stream: TStream): TOneOfOptionTwo; static;
        // Write OneOfOptionTwo to output stream
        procedure WriteTo(stream: TStream); override;
        function ToString: ansistring; override;
    end;

implementation

class function TOneOf.ReadFrom(stream: TStream): TOneOf;
var tag: Int32;
begin
    tag := stream.ReadInt32;
    case tag of
        0: result := TOneOfOptionOne.ReadFrom(stream);
        1: result := TOneOfOptionTwo.ReadFrom(stream);
        else raise Exception.Create('Unexpected tag value');
    end;
end;

constructor TOneOfOptionOne.Create(vecInt32: TArray<Int32>; longInt: Int64);
begin
    self.vecInt32 := vecInt32;
    self.longInt := longInt;
end;

class function TOneOfOptionOne.ReadFrom(stream: TStream): TOneOfOptionOne;
var longInt: Int64;
var vecInt32: TArray<Int32>;
var vecInt32Element: Int32;
var vecInt32Index: Int32;
begin
    vecInt32 := TArray<Int32>.Create;
    SetLength(vecInt32, stream.ReadInt32);
    for vecInt32Index := 0 to Length(vecInt32) - 1 do begin
        vecInt32Element := stream.ReadInt32;
        vecInt32[vecInt32Index] := vecInt32Element;
    end;
    longInt := stream.ReadInt64;
    result := TOneOfOptionOne.Create(vecInt32, longInt);
end;

procedure TOneOfOptionOne.WriteTo(stream: TStream);
var vecInt32Element: Int32;
begin
    stream.WriteInt32(0);
    stream.WriteInt32(Length(vecInt32));
    for vecInt32Element in vecInt32 do begin
        stream.WriteInt32(vecInt32Element);
    end;
    stream.WriteInt64(longInt);
end;

function TOneOfOptionOne.ToString: ansistring;
var vecInt32Element: Int32;
var vecInt32Index: Int32;
begin
    result := 'OptionOne {';
    result += 'vecInt32=';
    result += '[';
    for vecInt32Index := 0 to Length(vecInt32) - 1 do begin
        if vecInt32Index <> 0 then
            result += ', ';
        vecInt32Element := vecInt32[vecInt32Index];
        result += IntToStr(vecInt32Element);;
    end;
    result += ']';
    result += ', ';  
    result += 'longInt=';
    result += IntToStr(longInt);
    result += '}';
end;

constructor TOneOfOptionTwo.Create(value: Int32);
begin
    self.value := value;
end;

class function TOneOfOptionTwo.ReadFrom(stream: TStream): TOneOfOptionTwo;
var value: Int32;
begin
    value := stream.ReadInt32;
    result := TOneOfOptionTwo.Create(value);
end;

procedure TOneOfOptionTwo.WriteTo(stream: TStream);
begin
    stream.WriteInt32(1);
    stream.WriteInt32(value);
end;

function TOneOfOptionTwo.ToString: ansistring;
begin
    result := 'OptionTwo {';
    result += 'value=';
    result += IntToStr(value);
    result += '}';
end;

end.